rfcs/rdm-0079-collections.md

```diff
@@ -0,0 +1,337 @@
+
+This solution proposes collections to be modelled as generic trees, whose information is persisted in a database. 
+    
+That said, collections data model will be based in the "Materialized Path" pattern. Read this [documentation from MongoDB](https://www.mongodb.com/docs/manual/tutorial/model-tree-structures-with-materialized-paths/) for a good overview of this pattern. The main concept is that each node (row in the DB) stores its `path` from the root to itself, excluding. For our example, it looks as follows:
+
+
+In our proposed model, however, a unique `id` is to be stored in the `path` so the ascendants of a node can be derived from the path and done with primary key lookups:
+    
+```jsonld
+   { _id: 1, title: "Natural Sciences", path: "," },
+   { _id: 2, title: "Physical science", path: ",1," },
+   { _id: 3, title: "Mathematics", path: ",1," },
+   { _id: 4, title: "Particles", path: ",1,2," },
+```
+
+**Database tables**
+
+We identified three new tables for this design:
+                                     
+1) `Collection`
+2) `CollectionTree`
+3) `CommunityCollectionTree`
+                                     
+![](0079/6.png)
+
+In our previous example, this is the final model and entities with the highlighted path of some collections (in red): 
+                                     
+![](0079/7.png)
+
+                                     
+### Design considerations 
+                                     
+The idea behind splitting `CollectionTree` from `Collection`, is to be able to define "root" collections that do not hold any query and simply group collections under a semantic umbrella. In our previous example, it would be modelled as such:
+                                     
+```
+European Open Repository  <- Community
+    Subjects                  <- CollectionTree
+        Natural sciences          <- Collection
+        Physical sciences         <- Collection
+```
+
+An important aspect of this design is that the `CommunityCollectionTree` allows the `community_id` to be set to `null`, indicating that the collection can be treated as a `root` collection. This enables the retrieval of all records from the repository without restricting them to a specific community.
+                                     
+<a name="ctrees">🔑</a> All the collections must be linked to a `CollectionTree` for organization purposes, however the display of the `CollectionTree` `title` on the UI could be optional. In a future iteration one could add a column to control the `CollectionTree` visibility.
+                                     
+<a name="indices">🔑</a> Apart from the usual PK, FK indices, the following indices are to be added to increase the performance of the most common used queries:
+                                     
+1) Index on `Collection.path`, possibly containing the `depth` column to allow matching only subcollections until a certain depth. 
```

Comment:
do we define any requirements about the allowed depth? I remember we mentioned this during the discussion

Reply:
We had the depth in mind when modelling collections. This model (materialized paths) works with any depth, and we added the column to speed up queries for `depth > 1`. Any requirement / restrictions can then be added at the API or service level. 

As a side note, there's an optimized case, which is when `depth=1` (e.g. you want to query only the children of a node). The query is much faster since it does an exact text match:

```python
# Foo is the root node
foo = Collection("foo", parent=None)

# The children of 'foo' have the exact following path:
foo_children_path = ",foo,"
```
This can be used in DB like:

```sql
SELECT * collections
WHERE path = ",foo,"
```